### PR TITLE
[FIX] Melodic descriptor

### PR DESCRIPTION
**Pull request by dariusvalevicius**
Adjusted Melodic descriptor to run successfully on single NifTI input.
- Uses correct command ("melodic" and not "Melodic")
- Output directory correctly configured

Also added input and output types.

Bonus: If optional mask file input is not provided, the descriptor will call _bet_ to generate one from the 4D input file.

### CODE REVIEW

**Comment by prioux** [comment]
You could have noticed the newline problem by running bosh exec simulate:

```shell
unix% bosh exec simulate Melodic_fsl.json 
Generated Command:
if [ ]; then export MASKFILE=[MASK]; else echo '
Generating mask from input file...
' && bet f_infile_47 mask.nii.gz && export MASKFILE='mask.nii.gz'; fi; echo '
Running Melodic...
'; melodic -i f_infile_47 -o melodic_output -m $MASKFILE
```

**Comment by prioux** [comment]
3) use double quotes around your variable interpolation in the command

**Comment by dariusvalevicius** [comment]
I added an escape character to the newlines (a little cleaner than double quotes?)

bosh exec simulate output:
`Generated Command:
if [ ]; then export MASKFILE=[MASK]; else echo '\nGenerating mask from input file...\n' && bet f_infile_33.csv mask.nii.gz && export MASKFILE='mask.nii.gz'; fi; echo '\nRunning Melodic...\n'; melodic -i f_infile_33.csv -o melodic_output -m $MASKFILE`

**Comment by dariusvalevicius** [comment]
The MASKFILE environment variable is necessary given the option to provide a mask file or not, at least with the way I handled the logic (maybe there's a better way?).

**Comment by prioux** [comment]
Your echo command will print the `\n` literally, because echo doesn't support special sequences like `\n`. Try your code first, it's easy to see in a shell that it doesn't work:

```
> echo '\nHello\n'
\nHello\n
```

Also, I wasn't telling you not to use a MASKFILE variable, I was telling you not to export it to the environment unless necessary. Environment variables and shell variables have two namespaces.

**Comment by prioux** [comment]
```
Shell variable:

unix> blah=hello
unix> echo $blah
hello
unix> env | grep blah
(nothing)

Env variable:

unix> export glop=bye
unix> echo $glop
bye
unix> env | grep glop
glop=bye
```

**Comment by dariusvalevicius** [comment]
I swapped the env variable for a regular variable. For the newline, I swapped echo for printf and kept the escape characters in front of the newline in JSON and this produces the correct behaviour both with bosh exec simulate and running the actual command.

File: userfiles/melodic_output/melodic_output.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# A FileCollection to model the output of a oxford_asl
```
Comment (natacha-beck) [inline userfiles/melodic_output/melodic_output.rb:23]: To model and output of Melodic instead of oxford_asl.

**Review by natacha-beck — APPROVED** [review]

**Review by natacha-beck — APPROVED** [review]

**Comment by prioux** [comment]
In CBRAIN we already have a data model called `FslMelodicOutput`, why do we now need a `MelodicOutput` ? Are they different?

**Comment by prioux** [comment]
Can we just use the existing FslMelodicOutput class instead?

**Comment by prioux** [comment]
I just checked, in the production CBRAIN there have never been a single FslMelodicOutput created anyway, so I doubt there is a compatibility problem there.

**Comment by dariusvalevicius** [comment]
I did not notice the existing FSL Melodic filetype, it looks like it is configured for the old ruby version of Melodic (as far as I remember, not functional when I tested it, which would explain why no files exist).

The old output fsl_melodic_output.rb file seems to have a lot of extra complexity (200 lines of code), so it would likely break with the new boutiques version?

**Comment by prioux** [comment]
Right so the viewer code in the old FslMelodic starts with this big block of warning:

```
99e13b0c Tristan  2015-06-26   ##################################################################################################
99e13b0c Tristan  2015-06-26   # This method returns the content of an HTML file modified so that it can be                   
99e13b0c Tristan  2015-06-26   # displayed in the content section of the userfile page.  It                                   
99e13b0c Tristan  2015-06-26   # basically removes some HTML tags (html, body, object, head, title)                           
99e13b0c Tristan  2015-06-26   # and tweaks the URLs found in <a> and <img> tags.  It is quite                                
99e13b0c Tristan  2015-06-26   # specific to the HTML pages generated by FSL melodic. A more                                  
99e13b0c Tristan  2015-06-26   # generic implementation would require a complete parsing of the                               
99e13b0c Tristan  2015-06-26   # HTML code.  It is recursively called for each iframe contained in                            
99e13b0c Tristan  2015-06-26   # the page.                                                                                    
99e13b0c Tristan  2015-06-26   # Parameters:                                                                                  
99e13b0c Tristan  2015-06-26   #  * file_path: the path of the HTML file whose content will be read, modified and returned.   
99e13b0c Tristan  2015-06-26   #  * dir_name: the directory name of the HTML file in the melodic file collection.             
99e13b0c Tristan  2015-06-26   ###############################################################################################
```

I will assume this is no longer usable, 7 years later, and that modern melodic structures its output better.

**Comment by natacha-beck** [comment]
It has a viewer indeed.

**Comment by natacha-beck** [comment]
+ We have now the SubViewer mechanism.

**Pull request merged**